multinet/user.py

```diff
@@ -0,0 +1,98 @@
+"""User data and functions."""
+
+from uuid import uuid4
+from arango.collection import StandardCollection
+
+from multinet.db import db
+from multinet.auth.types import GoogleUserInfo, UserInfo, User
+
+from typing import Optional, Dict
+
+
+# This essentially duplicated the UserInfo TypedDict.
+# Is this a good reason to switch to dataclasses?
+USER_FIELDS = {"email", "name", "family_name", "given_name", "picture", "sub"}
+
+
+def user_collection() -> StandardCollection:
+    """Return the collection that contains user documents."""
+    sysdb = db("_system")
+
+    if not sysdb.has_collection("users"):
+        sysdb.create_collection("users")
+
+    return sysdb.collection("users")
+
+
+def user_exists(userinfo: GoogleUserInfo) -> bool:
+    """Return the existance of a user."""
+    return load_user(userinfo) is not None
+
+
+def load_user(userinfo: UserInfo) -> Optional[User]:
+    """Return a user doc if it exists, else None."""
+    coll = user_collection()
+    user = list(coll.find({"sub": userinfo["sub"]}, limit=1))
+
+    if not user:
+        return None
+
+    return user[0]
+
+
+def save_user(user: User) -> User:
+    """Update a user using the provided user object."""
+    coll = user_collection()
+    inserted_info: Dict = coll.update(user)  # type: ignore
+    return next(coll.find({"_id": inserted_info["_id"]}, limit=1))
+
+
+def register_user(userinfo: UserInfo) -> User:
+    """Register a user with the given user info."""
+    coll = user_collection()
+
+    # TODO: Fix mypy errors
+    document: User = {**userinfo, "multinet": {}}
+    inserted_info: Dict = coll.insert(document)  # type: ignore
+
+    return next(coll.find(inserted_info, limit=1))
+
+
+def set_user_cookie(user: User) -> User:
+    """Update the user cookie."""
+    new_user: User = dict(user)
+
+    new_cookie = uuid4().hex
+    new_user["multinet"]["session"] = new_cookie
```

Comment:
Is there any chance of a collision here?

Reply:
Collision as in two users having the same session cookie? Not really, according to the [wikipedia page](https://en.wikipedia.org/wiki/Universally_unique_identifier), there are 5.3 undecillion possible UUID4 values, with the following anecdote:

> the probability to find a duplicate within 103 trillion version-4 UUIDs is one in a billion.

So we're probably fine.

Reply:
Works for me, figured I'd check